### PR TITLE
[SYCL][Graph] Add support for CUDA backend

### PR DESCRIPTION
**Pull request by mfrancepillois**
Adds support for enabling CUDA backend with Sycl-Graph:

- Enables Graph e2e tests with CUDA backend.

Companion PR with CUDA UR adapter changes is here: https://github.com/oneapi-src/unified-runtime/pull/932

## Authors
Co-authored-by: Maxime France-Pillois  <maxime.francepillois@codeplay.com>
Co-authored-by: Ben Tracy <ben.tracy@codeplay.com>
Co-authored-by: Ewan Crawford <ewan@codeplay.com>

### CODE REVIEW

**Comment by Bensuo** [comment]
CI fails appear to be unrelated, as seen in other PRs, for example here: https://github.com/intel/llvm/pull/11553#issuecomment-1766717083

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 1]
```diff
@@ -53,14 +53,16 @@ endif()
...
+  set(UNIFIED_RUNTIME_REPO "https://github.com/bensuo/unified-runtime.git")
+
+  # commit 74f42f8fdd3b92b355c5769bd13d4d982e839c78
+  # Author: Ben Tracy <ben.tracy@codeplay.com>
+  # Date:   Fri Oct 6 12:57:03 2023 +0100
+  # [CUDA][EXP] CUDA adapter support for command buffers
+  # - Implement initial subset of command buffer functions for CUDA
+  # - Report command buffer extension as supported from device query
+  # - Some additional methods and refactoring to reuse existing adapter code
+  set(UNIFIED_RUNTIME_TAG 5ecaf66a2d0325cb9b8cfcb8e39a60810fa46cc2)
```
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:65]: Now that https://github.com/oneapi-src/unified-runtime/pull/932 has merged please update `UNIFIED_RUNTIME_TAG` as follows and pull in the latest sycl branch changes. This should then be ready to take out of draft.

```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit be53fb3ba5bf1ac33051456d72c61b6a01d94a72
  # Merge: 55d432c 5ecaf66
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Tue Oct 31 13:22:52 2023 +0000
  #     Merge pull request #932 from Bensuo/cuda-cmd-buffers
  #     [CUDA][EXP] CUDA adapter support for command buffers
```

**Review by kbenzie — CHANGES_REQUESTED** [review]

**Comment by EwanC** [comment]
Could we double check this covers all the Graphs E2E tests that've been added since this PR was opened, e.g.
* https://github.com/intel/llvm/pull/11418
* https://github.com/intel/llvm/pull/11505/

**Review by kbenzie — APPROVED** [review]
UR LGTM

**Review by kbenzie — APPROVED** [review]
UR LGTM

File: sycl/doc/design/CommandGraph.md [hunk 2]
```diff
@@ -221,3 +221,28 @@ Level Zero:
...
+The SYCL Graph CUDA backend relies on the
+[CUDA Graphs feature](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs),
+which is the CUDA public API for batching series of operations,
+such as kernel launches, connected by dependencies.
+
+UR commands (e.g. kernels) are mapped as graph nodes using the
+[CUDA Driver API](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__GRAPH.html#group__CUDA__GRAPH).
+The CUDA Driver API is preferred over the CUDA Runtime API to implement
+the SYCL Graph backend to remain consistent with other UR functions.
+Synchonizations between commands (UR sync-points) are implemented
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md:236]:
```suggestion
Synchronization between commands (UR sync-points) is implemented by
```

File: sycl/doc/design/CommandGraph.md [hunk 3]
```diff
@@ -221,3 +221,28 @@ Level Zero:
...
+using graph dependencies.
+
+Executable CUDA Graphs can be submitted to a CUDA stream
+in the same way as regular kernels.
+The CUDA backend enables enqueuing events to wait for into a stream.
+It also allows signaling the completion of a submission with an event.
+Therefore, submitting a UR command-buffer consists only of submitting to a stream
+the executable CUDA Graph that represent this series of operations.
+
+An executable CUDA Graph, which contains all commands and synchonization
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md:246]:
```suggestion
An executable CUDA Graph, which contains all commands and synchronization
```

File: sycl/doc/design/CommandGraph.md [hunk 4]
```diff
@@ -221,3 +221,28 @@ Level Zero:
...
+
+The SYCL Graph CUDA backend relies on the
+[CUDA Graphs feature](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs),
+which is the CUDA public API for batching series of operations,
+such as kernel launches, connected by dependencies.
+
+UR commands (e.g. kernels) are mapped as graph nodes using the
+[CUDA Driver API](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__GRAPH.html#group__CUDA__GRAPH).
+The CUDA Driver API is preferred over the CUDA Runtime API to implement
+the SYCL Graph backend to remain consistent with other UR functions.
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md:235]: Can we add a brief statement, why that is the case?

File: sycl/doc/design/images/SYCL-Graph-Architecture.svg
Comment (reble) [inline sycl/doc/design/images/SYCL-Graph-Architecture.svg:1]: Let's make the HW specific layer consistent and add Intel to the CPU, GPU, FPGA, ... box.

**Review by Bensuo — APPROVED** [review]

**Comment by kbenzie** [comment]
@intel/llvm-reviewers-runtime please review ASAP, this is blocking other UR changes.

**Comment by EwanC** [comment]
@intel/llvm-reviewers-runtime Could you give this PR a look please

**Review by steffenlarsen — APPROVED** [review]
Sorry, this one flew under my radar! LGTM!

**Pull request merged**